kubernetes: 1.5.6 -> 1.7.1

matejc opened the pull request:
done:

- addon manager 
- dns as addon (kubedns is not available in the kubernetes pkg anymore)
- dashboard addon
- node taints (to support unscedulable, which is true on master-only nodes)
- add storage backend option
- add RBAC
- flannel integration
- basic singlenode/multinode tests

TODO:
- per service kubelet config (waiting for pr by @offlinehacker)
- better cert generation using cfssl(waiting for pr by @offlinehacker)
- support for seeding of docker images(dockerPullImage)
- e2e tests

Note: deployed on @gatehubnet test cluster on hetzner, so far work ok

offlinehacker committed:
kubernetes module: authorization improvements

matejc committed:
kubernetes: 1.5.6 -> 1.6.4

matejc committed:
kubernetes: update service

matejc commented:
also..
I have 1.6.2 already deployed on my cluster of 3 servers with nixops, but without many features that I listed above

and single node configuration:
```
    services.kubernetes.roles = ["master" "node"];
    services.kubernetes.apiserver.storageBackend = "etcd3";
    services.kubernetes.apiserver.securePort = 4443;
    services.kubernetes.apiserver.port = 4480;
    services.kubernetes.apiserver.extraOpts = "--v=2";
    services.kubernetes.kubelet.hostname = "192.168.88.79";
    services.kubernetes.dns.serverIp = "10.10.10.11";
    services.kubernetes.dashboard.enable = true;
    services.dnsmasq.enable = true;
    services.dnsmasq.servers = ["/${config.services.kubernetes.dns.domain}/${config.services.kubernetes.dns.serverIp}#53"];
    virtualisation.docker.extraOptions = "--iptables=false --ip-masq=false -b cbr0";
    networking.bridges.cbr0.interfaces = [];
    networking.interfaces.cbr0 = {};
    networking.firewall.trustedInterfaces = ["cbr0"];
    networking.nat.internalInterfaces = ["cbr0"];
    networking.nat.externalInterface = "wlp3s0";
```

oh, and I almost forgot, any help is appreciated, specially for testing

offlinehacker commented:
Please don't make another directory, but make kubernetes a directory and put add-ons there, make kubernetes.nix into default.nix

offlinehacker commented:
Besides that I'm ok with pull request, and thank you, I will probably fix some things here and there in the future.

matejc committed:
kube-dns: 1.14.1

matejc committed:
kubernetes: add tests

NeQuissimus commented on nixos/modules/services/cluster/kubernetes/default.nix:
```diff
@@ -712,16 +712,23 @@ in {
 
       domain = mkOption  {
-        description = "Kubernetes dns domain under which to create names.";
+        description = "Kuberntes dns domain under which to create names.";
```
Kubernetes

NeQuissimus commented on nixos/modules/services/cluster/kubernetes/default.nix:
```diff
@@ -712,16 +712,23 @@ in {
 
     dns = {
-      enable = mkEnableOption "Kubernetes dns addon.";
+      enable = mkEnableOption "kubernetes dns service.";
```
Can you upper-case to "Kubernetes" and "DNS" everywhere?

matejc replied:
ok, will do

copumpkin commented on nixos/modules/services/cluster/kubernetes/kube-addon-manager.nix:
```diff
@@ -0,0 +1,54 @@
+            "containers" = [{
+                "command" = ["/bin/bash"
+                    "-c"
+                    "/opt/kube-addons.sh | tee /var/log/kube-addon-manager.log"
```
what's with /opt here?

matejc replied:
this is an addon manager which is a Pod, /opt/kube-addons.sh is from inside the container, its just the way this addon manager is ran.

https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-addons/kube-addon-manager.yaml

copumpkin commented on nixos/modules/services/cluster/kubernetes/dns.nix:
```diff
@@ -0,0 +1,240 @@
+{ cfg }: {
```
Are these files generated somehow? If so, it might be better to include the generator or instructions for how to regenerate them.

matejc replied:
they are generated, yaml -> json -> nix
yaml files are from here
https://github.com/kubernetes/kubernetes/tree/master/cluster/addons

will add instructions...

copumpkin replied:
If they're fully generated and this is in a module, it might make sense to just import them as straight JSON, possibly even referring to the external file? I don't think import-from-derivation is that big of a deal in system configurations.

matejc replied:
JSON has problems that it cannot have comments and extra syntax check of file is a big plus, as some files need to be changed a bit after generation, at least ip and port of the apiserver

I had a derivation of JSON files for each addon when I first started with this addons, but the passing of configuration options to the derivation itself.. its just a little dirty solution to rebuild addons every time configuration changes, at least for me

copumpkin commented on pkgs/applications/networking/cluster/kubernetes/dns.nix:
```diff
@@ -0,0 +1,40 @@
+    mkdir $GOPATH/src/k8s.io/dns
+    ln -s $(pwd)/cmd $GOPATH/src/k8s.io/dns/cmd
+    ln -s $(pwd)/pkg $GOPATH/src/k8s.io/dns/pkg
+    go build -o kube-dns ./cmd/kube-dns
```
Are you not able to use the standard Go builder stuff here? All this `GOPATH` manipulation should be handled for you if you use it.

matejc replied:
yeah, I tried with build go package function, and after a while I gave up, it did not found any of the dependencies, also tried with go2nix, got 404 errors for dependencies

matejc replied:
besides, I do not want to compile whole package, but just cmd/kube-dns

copumpkin replied:
Ah okay, in that case I'd just leave a comment saying that so the next reader doesn't start from 0 if they want to try to convert it to use the standard stuff. I know the go builder does have the option to build only a subset of the targets. cc @kamilchm

matejc commented:
thanks all for comments, keep 'em comin'

... now, that pesky RBAC is next

matejc commented:
now it is time for deploying this to actual test clusters, if anyone wants to try out, now it would be the time, please feel free to ask any questions or write comments

I will upgrade my personal cluster with this update, but it might take me a week to find and reserve time to do that, so again, please do try it out before we merge this

benley commented:
fyi, kubernetes 1.6.3 has since been released

matejc commented:
thanks, will update

On May 19, 2017 05:24, "Benjamin Staffin" <notifications@github.com> wrote:

> fyi, kubernetes 1.6.3 has since been released
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/25426#issuecomment-302599018>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA0K8uKieBQeiW-emCF1nf9Nko3M7LB1ks5r7QtqgaJpZM4NNtzW>
> .
>

matejc commented:
I have upgraded kubernetes to 1.6.4, and added DNS test, because when I upgraded my personal cluster and enabled kubernetes.dns.enable option, the containers had no access to dns server on host, I also fixed that with extra iptables rule. Everything else went smooth during upgrade procedure.

matejc commented:
OK, rebased to master and tidied up my commits, this is getting ready for merge, any last requests or comments?

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -56,6 +53,35 @@ let
+      effect = mkOption {
+        description = "Effect of taint.";
+        example = "NoSchedule";
+        type = types.str;
```
yup

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -175,6 +201,14 @@ in {
+        description = ''
+          Kubernetes apiserver storage backend.
+        '';
+        default = "etcd2";
```
yes

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -536,11 +641,29 @@ in {
         default = {};
       };
 
+      applyManifests = mkOption {
```
I will add a better description

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -536,11 +641,29 @@ in {
+        type = types.bool;
+      };
+
+      unschedulable = mkOption {
```
:)

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -536,11 +641,29 @@ in {
+      };
+
+      taints = mkOption {
+        description = ".";
```
woops, yes will add

matejc commented on nixos/modules/services/cluster/kubernetes.nix:
```diff
@@ -599,7 +752,7 @@ in {
         wantedBy = [ "kubernetes.target" ];
         after = [ "network.target" "docker.service" "kube-apiserver.service" ];
-        path = with pkgs; [ gitMinimal openssh docker utillinux iproute ethtool thin-provisioning-tools iptables ];
+        path = with pkgs; [ gitMinimal openssh docker utillinux iproute ethtool thin-provisioning-tools iptables ] ++ cfg.path;
```
yes, that and also ceph binary for distributed storage solution (rook.io), but that is for the next PR

matejc commented on nixos/tests/kubernetes/rbac.nix:
```diff
@@ -0,0 +1,206 @@
+            virtualisation.diskSize = 4096;
+            networking.interfaces.eth1.ip4 = mkForce [{address = servers.master; prefixLength = 24;}];
+            networking.primaryIPAddress = mkForce servers.master;
+            services.kubernetes.apiserver.authorizationMode = mkForce ["ABAC" "RBAC"];
```
ok

matejc committed:
kubernetes: fix minor issues

offlinehacker committed:
kubernetes module: flannel support, minor fixes

- add flannel support
- remove deprecated authorizationRBACSuperAdmin option
- rename from deprecated poratalNet to serviceClusterIpRange
- add nodeIp option for kubelet
- kubelet, add br_netfilter to kernelModules
- enable firewall by default
- enable dns by default on node and on master
- disable iptables for docker by default on nodes
- dns, restart on failure
- update tests

and other minor changes

matejc commented:
this PR is still in WIP state, @offlinehacker has refactored it, and DNS service will be packaged as addon again, because of its tight integration to kubernetes, I have already done this, but have to finish it, but its good that we have tests now \o/

offlinehacker committed:
kubernetes module: per service kubeconfig support

offlinehacker committed:
kubernetes module: add support for common CA file

offlinehacker committed:
kubernetes module: default auth mode to only RBAC

offlinehacker committed:
kubernetes module: fix cidr ranges

offlinehacker committed:
kubernetes module: add storage and tolerations addmission controllers

offlinehacker committed:
kubernetes module: enable leader elect by default

offlinehacker commented:
I have one additional change regarding proper cert generation, RBAC only mode and support for per service kubeconfig files and other smaller fixes, waiting on my unstable branch to be merged in here when I make pull request to matejc. These changes and DNS add-on by matejc would bring featureset the same as official kubernetes setup. Later e2e test integration follows in other pull requests.

offlinehacker committed:
kubernetes module: kubelet, add socat to path for kubectl portforward

sweber83 commented:
We tested your pull request with the vSphere cloud provider and stumbled onto a small issue.
For the vSphere volume to work, the kubelet needs to exec mkfs, which it can't with your config.
This is easily fixed:
```
systemd.services.kubelet.path = [pkgs.e2fsprogs];
```

moretea commented:
@offlinehacker do you have that code available? Would love to help you out, if possible. I also noticed that (at least) the one-node nixos test is failing.

matejc commented:
This PR is on my mind every day. 

Progress update: @offlinehacker is currently working on updating kubernetes to 1.7 version, some tests still need fixing and DNS is fixed as far as I know, he will create a pull request to this branch on my nixpkgs and I will merge when all tests will pass on my end too. @offlinehacker is quite busy and is somewhat a perfectionist :)  so this pull request will be top notch once finished

@sweber83 I need that feature too, so it will be in final version of this PR

srhb commented:
@matejc Thanks for the update! I was working separately on 1.7 from the old source distribution since that seemed a bit simpler to understand, but it's great to know that this is progressing. Please do tell if we can do anything to help. :-)

NeQuissimus commented:
How is this looking?

matejc commented:
Yeah @offlinehacker and I, are quite busy.

@offlinehacker can you make a PR on my branch with everything you have done, even if it is not finished? That way, if not me, then somebody else can finish the PR.

offlinehacker committed:
kubernetes: 1.6.4 -> 1.7.1

offlinehacker committed:
kuberentes: expose kube-addons script

offlinehacker committed:
kubernetes module: add featureGates option

offlinehacker committed:
kubernetes module: fix documentation links

offlinehacker committed:
kubernetes module: webhook authorization for kubelet

offlinehacker committed:
kubernetes module: seedDockerImages option for seeding docker images built with nix

offlinehacker committed:
kubernetes module: refactor module system, kube-dns as module

offlinehacker committed:
cni: 0.3.0 -> 0.5.2

offlinehacker commented:
Ok, so i updated this. It's a setup we use in one of our compute clusters on hetzner. It's still untested on master, but I will update tests in following days, as we need to update our cluster too, and i need to make sure it will work after update. In our setup everything we need(and we are quite power-kubernetes user) works:

- Basic kubernetes setup(everything seems to work)
- Flannel networking with CNI
- RBAC auth, with kube-apiserver and kubelet rbac authenticated
- Addon manager
- DNS as addon
- Kubernetes dashboard as addon

This is nixos config we currently are using. It's a single kubernetes, single etcd node cluster with flannel networking, with RBAC auth.

```
    networking = {
      domain = "myawesomecluster.cluster.yourdomain.net";

      firewall = {
        allowedTCPPorts = [
          10250 # kubelet
          2379 2380  # etcd
          443 # kubernetes apiserver
        ];
        trustedInterfaces = ["docker0"];

        # allow any traffic from all of the nodes in the cluster
        extraCommands = lib.concatMapStrings  (node: ''
          iptables -A INPUT -s ${node.config.networking.publicIPv4} -j ACCEPT
        '') (lib.attrValues nodes);
      };
    };

    services.kubernetes = {
      featureGates = ["AllAlpha"];

      caFile = "/var/lib/kubernetes/certs/ca.pem";
      verbose = true;

      apiserver = {
        tlsCertFile = "/var/lib/kubernetes/certs/kube-apiserver.pem";
        tlsKeyFile = "/var/lib/kubernetes/certs/kube-apiserver-key.pem";
        kubeletClientCertFile = "/var/lib/kubernetes/certs/kubelet-client.pem";
        kubeletClientKeyFile = "/var/lib/kubernetes/certs/kubelet-client-key.pem";
        serviceAccountKeyFile = "/var/lib/kubernetes/certs/kube-service-accounts.pem";
      };
      etcd = {
        servers = ["https://etcd.${config.networking.domain}:2379"];
        certFile = "/var/lib/kubernetes/certs/etcd-client.pem";
        keyFile = "/var/lib/kubernetes/certs/etcd-client-key.pem";
      };
      kubeconfig = {
        server = "https://api.${config.networking.domain}";
      };
      kubelet = {
        tlsCertFile = "/var/lib/kubernetes/certs/kubelet.pem";
        tlsKeyFile = "/var/lib/kubernetes/certs/kubelet-key.pem";
        hostname = "${config.networking.hostName}.${config.networking.domain}";
        kubeconfig = {
          certFile = "/var/lib/kubernetes/certs/apiserver-client-kubelet.pem";
          keyFile = "/var/lib/kubernetes/certs/apiserver-client-kubelet-key.pem";
        };
      };
      controllerManager = {
        serviceAccountKeyFile = "/var/lib/kubernetes/certs/kube-service-accounts-key.pem";
        kubeconfig = {
          certFile = "/var/lib/kubernetes/certs/apiserver-client-kube-controller-manager.pem";
          keyFile = "/var/lib/kubernetes/certs/apiserver-client-kube-controller-manager-key.pem";
        };
      };
      scheduler = {
        kubeconfig = {
          certFile = "/var/lib/kubernetes/certs/apiserver-client-kube-scheduler.pem";
          keyFile = "/var/lib/kubernetes/certs/apiserver-client-kube-scheduler-key.pem";
        };
      };
      proxy = {
        kubeconfig = {
          certFile = "/var/lib/kubernetes/certs/apiserver-client-kube-proxy.pem";
          keyFile = "/var/lib/kubernetes/certs/apiserver-client-kube-proxy-key.pem";
        };
      };
      flannel.enable = true;
      addons.dashboard.enable = true;
    };
```

This is how we generate certs:

```
{
  pkgs ? import <nixpkgs> {},
  internalDomain ? "cloud.yourdomain.net",
  externalDomain ? "myawesomecluster.cluster.yourdomain.net",
  serviceClusterIp ? "10.0.0.1"
}:
let
  runWithCFSSL = name: cmd:
    builtins.fromJSON (builtins.readFile (
      pkgs.runCommand "${name}-cfss.json" {
        buildInputs = [ pkgs.cfssl ];
      } "cfssl ${cmd} > $out"
    ));

  writeCFSSL = content:
    pkgs.runCommand content.name {
      buildInputs = [ pkgs.cfssl ];
    } ''
      mkdir -p $out
      cd $out
      cat ${writeFile content} | cfssljson -bare ${content.name}
    '';

  noCSR = content: pkgs.lib.filterAttrs (n: v: n != "csr") content;
  noKey = content: pkgs.lib.filterAttrs (n: v: n != "key") content;

  writeFile = content: pkgs.writeText "content" (
    if pkgs.lib.isAttrs content then builtins.toJSON content
    else toString content
  );

  createServingCertKey = { ca, cn, hosts? [], size ? 2048, name ? cn }:
    noCSR (
      (runWithCFSSL name "gencert -ca=${writeFile ca.cert} -ca-key=${writeFile ca.key} -profile=server -config=${writeFile ca.config} ${writeFile {
        CN = cn;
        hosts = hosts;
        key = { algo = "rsa"; inherit size; };
      }}") // { inherit name; }
    );

  createClientCertKey = { ca, cn, groups ? [], size ? 2048, name ? cn }:
    noCSR (
      (runWithCFSSL name "gencert -ca=${writeFile ca.cert} -ca-key=${writeFile ca.key} -profile=client -config=${writeFile ca.config} ${writeFile {
        CN = cn;
        names = map (group: {O = group;}) groups;
        hosts = [""];
        key = { algo = "rsa"; inherit size; };
      }}") // { inherit name; }
    );

  createSigningCertKey = { C ? "xx", ST ? "x", L ? "x", O ? "x", OU ? "x", CN ? "ca", emailAddress ? "x", expiry ? "43800h", size ? 2048, name ? CN }:
    (noCSR (runWithCFSSL CN "genkey -initca ${writeFile {
      key = { algo = "rsa"; inherit size; };
      names = [{ inherit C ST L O OU CN emailAddress; }];
    }}")) // {
      inherit name;
      config.signing = {
        default.expiry = expiry;
        profiles = {
          server = {
            inherit expiry;
            usages = [
              "signing"
              "key encipherment"
              "server auth"
            ];
          };
          client = {
            inherit expiry;
            usages = [
              "signing"
              "key encipherment"
              "client auth"
            ];
          };
          peer = {
            inherit expiry;
            usages = [
              "signing"
              "key encipherment"
              "server auth"
              "client auth"
            ];
          };
        };
      };
    };

  ca = createSigningCertKey {};

  kube-apiserver = createServingCertKey {
    inherit ca;
    cn = "kube-apiserver";
    hosts = ["kubernetes.default" "kubernetes.default.svc" "localhost" "api.${externalDomain}" serviceClusterIp];
  };

  kubelet = createServingCertKey {
    inherit ca;
    cn = "kubelet";
    hosts = ["*.${externalDomain}"];
  };

  service-accounts = createServingCertKey {
    inherit ca;
    cn = "kube-service-accounts";
  };

  etcd = createServingCertKey {
    inherit ca;
    cn = "etcd";
    hosts = ["etcd.${externalDomain}"];
  };

  etcd-client = createClientCertKey {
    inherit ca;
    cn = "etcd-client";
  };

  kubelet-client = createClientCertKey {
    inherit ca;
    cn = "kubelet-client";
    groups = ["system:masters"];
  };

  apiserver-client = {
    kubelet = createClientCertKey {
      inherit ca;
      cn = "apiserver-client-kubelet";
      groups = ["system:nodes"];
    };

    kube-proxy = createClientCertKey {
      inherit ca;
      name = "apiserver-client-kube-proxy";
      cn = "system:kube-proxy";
      groups = ["system:kube-proxy" "system:nodes"];
    };

    kube-controller-manager = createClientCertKey {
      inherit ca;
      name = "apiserver-client-kube-controller-manager";
      cn = "system:kube-controller-manager";
      groups = ["system:masters"];
    };

    kube-scheduler = createClientCertKey {
      inherit ca;
      name = "apiserver-client-kube-scheduler";
      cn = "system:kube-scheduler";
      groups = ["system:kube-scheduler"];
    };

    admin = createClientCertKey {
      inherit ca;
      cn = "admin";
      groups = ["system:masters"];
    };
  };
in {
  master = pkgs.buildEnv {
    name = "master-keys";
    paths = [
      (writeCFSSL (noKey ca))
      (writeCFSSL kube-apiserver)
      (writeCFSSL kubelet-client)
      (writeCFSSL apiserver-client.kube-controller-manager)
      (writeCFSSL apiserver-client.kube-scheduler)
      (writeCFSSL service-accounts)
      (writeCFSSL etcd)
    ];
  };

  worker = pkgs.buildEnv {
    name = "worker-keys";
    paths = [
      (writeCFSSL (noKey ca))
      (writeCFSSL kubelet)
      (writeCFSSL apiserver-client.kubelet)
      (writeCFSSL apiserver-client.kube-proxy)
      (writeCFSSL etcd-client)
    ];
  };

  admin = writeCFSSL apiserver-client.admin;
}
```

Make sure to generate certs on encrypted partition out of your main nix store(for example use proot) if you will use them on a production cluster.

Fresheyeball commented:
Fabulous! :+100:

offlinehacker commented:
What is left are nixos tests running kubernetes e2e functional test suite for different setups(single node, multinode,...). @matejc is working on that. After that we have complete and tested kubernetes integration in nixos.

Should i mention that we are the only ones that can do offline deploy of kubernetes, even addons, meaning we can deploy without docker registry and reliance on Google docker registries, and are only ones actually verifying hashes of third party(Google) docker images.

matejc commented:
I have fixed the DNS test, also gave kubernetes module a few fixes, thanks to @offlinehacker . Now I have working kubernetes nix function for testing virtually any kubernetes setup, right now the DNS test is working, e2e is running (but some tests are failing). Will push those two tests later today or tomorrow - I have to clean up the code a bit, hopefully someone else can help with writing various tests and specially fix e2e test itself which covers majority of cases.

matejc committed:
kubernetes: fix dns addon hashes, fix clusterDns, enable proxy on master

matejc committed:
kubernetes: build e2e.test

matejc committed:
kubernetes: fix tests

matejc commented:
Well... I guess I dreamed about writing the comment here because it did not happen'

I have made some small kubernetes test abstraction and added dns, e2e and rbac tests, note that when e2e will be adjusted, we will not need rbac test because e2e should take care of it along with a bunch of other cases.

Also, rebased to master at the time.

PS: travis-ci said: no space left on device

matejc commented:
This should be finally it! Now applause or throw tomatoes.

offlinehacker commented:
So while these tests are not perfect and are a bit complex, well at least now we have them. Thanks!

 So i was just wondering what's the smallest possible single node setup without all cert generation thingy, well that's something what most of nixos users will have, and we want to keep it simple. Will you do this @matejc?

After that is done, I'm all for LGTM, as we spent a lot of time on this, it's battle tested in production and except for e2e tests, it's close to perfect.

NeQuissimus commented:
LGTM!

NeQuissimus commented:
LGTM!

globin commented:
I'm getting different hashes for the dns images:
```
diff --git a/nixos/modules/services/cluster/kubernetes/dns.nix b/nixos/modules/services/cluster/kubernetes/dns.nix
index 743fbdedfc5..226fdadffd1 100644
--- a/nixos/modules/services/cluster/kubernetes/dns.nix
+++ b/nixos/modules/services/cluster/kubernetes/dns.nix
@@ -8,19 +8,19 @@ let
   k8s-dns-kube-dns = pkgs.dockerTools.pullImage {
     imageName = "gcr.io/google_containers/k8s-dns-kube-dns-amd64";
     imageTag = version;
-    sha256 = "1zq8xl0rac99r925xmalmx1xc3pgzprh6csrzck9k8fnqp50smdk";
+    sha256 = "0q97xfqrigrfjl2a9cxl5in619py0zv44gch09jm8gqjkxl80imp";
   };
 
   k8s-dns-dnsmasq-nanny = pkgs.dockerTools.pullImage {
     imageName = "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64";
     imageTag = version;
-    sha256 = "1dzw1qb97w2km6c5p564g1r5lnvyblwg5h2cssn09v7x4w3lf732";
+    sha256 = "051w5ca4qb88mwva4hbnh9xzlsvv7k1mbk3wz50lmig2mqrqqx6c";
   };
 
   k8s-dns-sidecar = pkgs.dockerTools.pullImage {
     imageName = "gcr.io/google_containers/k8s-dns-sidecar-amd64";
     imageTag = version;
-    sha256 = "1k9k6xklc1b74slpldnjsciikhkfhi0nl2n1vn1xr1y40dggssy4";
+    sha256 = "1z0d129bcm8i2cqq36x5jhnrv9hirj8c6kjrmdav8vgf7py78vsm";
   };
 
   cfg = config.services.kubernetes.addons.dns;
```

And `kubernetes.dns.port` does not exist => eval error in the multi-node test at least

matejc commented:
@globin looks like you are missing my last 3 commits ..

globin commented:
Ah yeah, missed updating my PR git remote, the e2e multinode test has been running for a while now, how long does it take on your machine?

globin commented:
ping @matejc, does the e2e test run on your machine?

offlinehacker commented:
@globin as far as i know there are still issues with e2e tests. This would be the hardest to implement, but would prove that all features of kubernetes on nixos work, also in the future.

offlinehacker commented:
@globin if you manage to fix e2e tests, this would be really awesome :)

globin commented:
Ok, I think we should merge nonetheless since this seems to work far better than what we currently have

NeQuissimus commented:
The e2e test just runs forever, it does not fail, does it?

offlinehacker commented:
@globin i guess you are right, it has almost been a year :-P First commit was end of 2016

offlinehacker commented:
Will this land in 17.09 release?

globin commented:
It is cherry-picked already and should be included in the current nixos-17.09 (beta) channel release already!

NeQuissimus commented:
Hydra is not running these tests, right?
Because I am getting what is mentioned in #29636 every time I try.

I tried teaching our `pullImage` / skopeo about certificates with ` --src-cert-dir='${pkgs.cacert}/etc/ssl/certs'` but that just seg-faults skopeo...

datakurre commented:
Thanks for all this! Yet, it would be very nice, if one of you would be able to document, how to get started with kubernets and nixops. I believe that I got a cluster running with nixops on virtualbox by quite strictly following this thread and new tests in 17.09, but I would get call it very convenient experience :)

voobscout commented:
@datakurre, mind sharing your configs for the cluster? Much appreciated!
